File: leetcode-js/046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let res = [[nums[0]]];

    for (let i = 1; i < nums.length; i++) {
        const newRes = [];

        for (const r of res) {
            for (let j = 0; j <= i; j++) {
                r.splice(j, 0, nums[i]);
                newRes.push(r.slice());
                r.splice(j, 1);
            }
        }

        res = newRes;
    }

    return res;
};

File: leetcode-java/046-permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        if (nums.length == 0) {
            return new ArrayList<>();
        }

        List<List<Integer>> res = new ArrayList<>();
        res.add(Arrays.asList(nums[0]));

        for (int i = 1; i < nums.length; i++) {
            List<List<Integer>> newRes = new ArrayList<>();
            for (List<Integer> r : res) {
                for (int j = 0; j <= i; j++) {
                    List<Integer> n = new ArrayList<>(r);
                    n.add(j, nums[i]);
                    newRes.add(n);
                }
            }
            res = newRes;
        }

        return res;
    }
}
